src/zkevm_specs/evm_circuit/execution/precompiles/ecpairing.py

```diff
@@ -0,0 +1,76 @@
+from zkevm_specs.evm_circuit.instruction import Instruction
+from zkevm_specs.evm_circuit.table import (
+    CallContextFieldTag,
+    EccOpTag,
+    FixedTableTag,
+)
+from zkevm_specs.util import FQ, Word, Bn254PairingBaseGas, Bn254PairingPerPointGas
+
+# Input is a multiple of 6 32-byte values, 6 * 32 = 192 (bytes)
+BYTES_PER_PAIRING = 192
+
+
+def ecPairing(instruction: Instruction):
+    is_success = instruction.call_context_lookup(CallContextFieldTag.IsSuccess)
+    calldata_len = instruction.call_context_lookup(CallContextFieldTag.CallDataLength)
+    address_word = instruction.call_context_lookup_word(CallContextFieldTag.CalleeAddress)
+    address = instruction.word_to_address(address_word)
+    instruction.fixed_lookup(
+        FixedTableTag.PrecompileInfo,
+        FQ(instruction.curr.execution_state),
+        address,
+        FQ(Bn254PairingBaseGas),
+    )
+
+    # Get input_rlc and result
+    input_rlc: FQ = instruction.curr.aux_data[0]
+    input_pairs: FQ = instruction.curr.aux_data[1]
+    is_valid_input: FQ = instruction.curr.aux_data[2]
+    output: FQ = instruction.curr.aux_data[3]
+
+    # output indicates whether the pairing is successful or not
+    instruction.constrain_equal(is_success.expr(), output)
```

Comment:
Should be `is_valid_input` here? It's successful call even the pairing result is `0`

Reply:
Valid inputs still could have failed pairing result (i.e. output == 0). Assuming p1 and p3 are valid on G1, and p2 and p4 are valid on G2, it still could cause e(p1, p2) != e(p3, p4) as you mentioned before. You can see [this test case](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/504/files#diff-278000e8769164199daab88f107a441a6db87852a1511ed351e5a2ebcefadd45R341-R370).  

I'm wondering what kinds of scenario will cuase a successful call but the pairing result is `0`?

Reply:
`is_success` is not equal to `return_data`, you can try this example and you'll see that `return_data = 0` but `is_success = 1`:

```
PUSH32 0x2CF44499D5D27BB186308B7AF7AF02AC5BC9EEB6A3D147C186B21FB1B76E18DA
PUSH1 0x00
MSTORE
PUSH32 0x2C0F001F52110CCFE69108924926E45F0B0C868DF0E7BDE1FE16D3242DC715F6
PUSH1 0x20
MSTORE
PUSH32 0x1FB19BB476F6B9E44E2A32234DA8212F61CD63919354BC06AEF31E3CFAFF3EBC
PUSH1 0x40
MSTORE
PUSH32 0x22606845FF186793914E03E21DF544C34FFE2F2F3504DE8A79D9159ECA2D98D9
PUSH1 0x60
MSTORE
PUSH32 0x2BD368E28381E8ECCB5FA81FC26CF3F048EEA9ABFDD85D7ED3AB3698D63E4F90
PUSH1 0x80
MSTORE
PUSH32 0x2FE02E47887507ADF0FF1743CBAC6BA291E66F59BE6BD763950BB16041A0A85E
PUSH1 0xa0
MSTORE
PUSH32 0x2CF44499D5D27BB186308B7AF7AF02AC5BC9EEB6A3D147C186B21FB1B76E18DA
PUSH1 0xc0
MSTORE
PUSH32 0x2C0F001F52110CCFE69108924926E45F0B0C868DF0E7BDE1FE16D3242DC715F6
PUSH1 0xe0
MSTORE
PUSH32 0x1971FF0471B09FA93CAAF13CBF443C1AEDE09CC4328F5A62AAD45F40EC133EB4
PUSH2 0x100
MSTORE
PUSH32 0x091058A3141822985733CBDDDFED0FD8D6C104E9E9EFF40BF5ABFEF9AB163BC7
PUSH2 0x120
MSTORE
PUSH32 0x2A23AF9A5CE2BA2796C1F4E453A370EB0AF8C212D9DC9ACD8FC02C2E907BAEA2
PUSH2 0x140
MSTORE
PUSH32 0x23A8EB0B0996252CB548A4487DA97B02422EBC0E834613F954DE6C7E0AFDC1FC
PUSH2 0x160
MSTORE
PUSH0
PUSH0
PUSH2 0x180
PUSH0
PUSH0
PUSH1 0x08
GAS
CALL
```

Reply:
I think I got your point (hopefully). 
This was fixed in [5978ddc](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/504/commits/5978ddca149e095dec0c338c249edf0254f48d12). There are two changes in [5978ddc](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/504/commits/5978ddca149e095dec0c338c249edf0254f48d12), one is you mentioned, "It's successful call even the pairing result is 0" and I added some comment, please help confirm my understanding is correct or not. The other is [here](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/504/commits/5978ddca149e095dec0c338c249edf0254f48d12#diff-c26e8a9134c8b18a978b04d77c1c2ae0b317d61ef6eb991ede314e771e3913b5R216), the assignment to `ecc_pairing_chip.output` should be lower part of output instead of `is_valid`.